File: 1386-cinema-seat-allocation/cinema-seat-allocation.cpp
class Solution {
public:
    int maxNumberOfFamilies(int n, vector<vector<int>>& reservedSeats) {
        unordered_map<int, set<int>> mp;
        for (auto reserve : reservedSeats) {
            int row = reserve[0];
            int col = reserve[1];
            mp[row].insert(col);
        }
        int result = 0;
        result += (n - mp.size()) * 2;
        for (auto &[row,book]:mp) {
            auto isAvailable=[&](int sear){
                return book.find(sear)==book.end();
            };
            bool GroupA = isAvailable(2) && isAvailable(3) && isAvailable(4) &&
                          isAvailable(5);
            bool GroupB = isAvailable(4) && isAvailable(5) && isAvailable(6) &&
                          isAvailable(7);
            bool GroupC = isAvailable(6) && isAvailable(7) && isAvailable(8) &&
                          isAvailable(9);
            
            if (GroupA && GroupC){
                result +=2;
            }
            else if(GroupA || GroupC || GroupB){
                result +=1;
            }
        }
        return result;
        
    }
};

File: 1386-cinema-seat-allocation/cinema-seat-allocation.py
class Solution(object):
    def maxNumberOfFamilies(self, n, reservedSeats):
        mp = {}

        for reserve in reservedSeats:
            row = reserve[0]
            col = reserve[1]

            if row not in mp:
                mp[row] = set()

            mp[row].add(col)

        result = 0

        # rows with no reserved seats
        result += (n - len(mp)) * 2

        for row, book in mp.items():

            def isAvailable(seat):
                return seat not in book

            GroupA = (
                isAvailable(2) and
                isAvailable(3) and
                isAvailable(4) and
                isAvailable(5)
            )

            GroupB = (
                isAvailable(4) and
                isAvailable(5) and
                isAvailable(6) and
                isAvailable(7)
            )

            GroupC = (
                isAvailable(6) and
                isAvailable(7) and
                isAvailable(8) and
                isAvailable(9)
            )

            if GroupA and GroupC:
                result += 2
            elif GroupA or GroupB or GroupC:
                result += 1

        return result
        """
        :type n: int
        :type reservedSeats: List[List[int]]
        :rtype: int
        """
        